File: java/test/Ice/exceptions/AMDThrowerI.java
// **********************************************************************
//
// Copyright (c) 2003-2009 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package test.Ice.exceptions;

import test.Ice.exceptionsAMD.Test.A;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_shutdown;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_supportsAssertException;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_supportsUndeclaredExceptions;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwAasA;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwAorDasAorD;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwAssertException;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwBasA;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwBasB;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwCasA;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwCasB;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwCasC;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwLocalException;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwNonIceException;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwUndeclaredA;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwUndeclaredB;
import test.Ice.exceptionsAMD.Test.AMD_Thrower_throwUndeclaredC;
import test.Ice.exceptionsAMD.Test.B;
import test.Ice.exceptionsAMD.Test.C;
import test.Ice.exceptionsAMD.Test.D;
import test.Ice.exceptionsAMD.Test._ThrowerDisp;

public final class AMDThrowerI extends _ThrowerDisp
{
    public
    AMDThrowerI(Ice.ObjectAdapter adapter)
    {
        _adapter = adapter;
    }

    public void
    shutdown_async(AMD_Thrower_shutdown cb, Ice.Current current)
    {
        _adapter.getCommunicator().shutdown();
        cb.ice_response();
    }

    public void
    supportsUndeclaredExceptions_async(AMD_Thrower_supportsUndeclaredExceptions cb, Ice.Current current)
    {
        cb.ice_response(true);
    }

    public void
    supportsAssertException_async(AMD_Thrower_supportsAssertException cb, Ice.Current current)
    {
        cb.ice_response(true);
    }

    public void
    throwAasA_async(AMD_Thrower_throwAasA cb, int a, Ice.Current current)
        throws A
    {
        A ex = new A();
        ex.aMem = a;
        cb.ice_exception(ex);
    }

    public void
    throwAorDasAorD_async(AMD_Thrower_throwAorDasAorD cb, int a, Ice.Current current)
        throws A,
               D
    {
        if(a > 0)
        {
            A ex = new A();
            ex.aMem = a;
            cb.ice_exception(ex);
        }
        else
        {
            D ex = new D();
            ex.dMem = a;
            cb.ice_exception(ex);
        }
    }

    public void
    throwBasA_async(AMD_Thrower_throwBasA cb, int a, int b, Ice.Current current)
        throws A
    {
        B ex = new B();
        ex.aMem = a;
        ex.bMem = b;
        throw ex;
        //cb.ice_exception(ex);
    }

    public void
    throwBasB_async(AMD_Thrower_throwBasB cb, int a, int b, Ice.Current current)
        throws B
    {
        B ex = new B();
        ex.aMem = a;
        ex.bMem = b;
        throw ex;
        //cb.ice_exception(ex);
    }

    public void
    throwCasA_async(AMD_Thrower_throwCasA cb, int a, int b, int c, Ice.Current current)
        throws A
    {
        C ex = new C();
        ex.aMem = a;
        ex.bMem = b;
        ex.cMem = c;
        cb.ice_exception(ex);
    }

    public void
    throwCasB_async(AMD_Thrower_throwCasB cb, int a, int b, int c, Ice.Current current)
        throws B
    {
        C ex = new C();
        ex.aMem = a;
        ex.bMem = b;
        ex.cMem = c;
        cb.ice_exception(ex);
    }

    public void
    throwCasC_async(AMD_Thrower_throwCasC cb, int a, int b, int c, Ice.Current current)
        throws C
    {
        C ex = new C();
        ex.aMem = a;
        ex.bMem = b;
        ex.cMem = c;
        cb.ice_exception(ex);
    }

    public void
    throwUndeclaredA_async(AMD_Thrower_throwUndeclaredA cb, int a, Ice.Current current)
    {
        A ex = new A();
        ex.aMem = a;
        cb.ice_exception(ex);
    }

    public void
    throwUndeclaredB_async(AMD_Thrower_throwUndeclaredB cb, int a, int b, Ice.Current current)
    {
        B ex = new B();
        ex.aMem = a;
        ex.bMem = b;
        cb.ice_exception(ex);
    }

    public void
    throwUndeclaredC_async(AMD_Thrower_throwUndeclaredC cb, int a, int b, int c, Ice.Current current)
    {
        C ex = new C();
        ex.aMem = a;
        ex.bMem = b;
        ex.cMem = c;
        cb.ice_exception(ex);
    }

    public void
    throwLocalException_async(AMD_Thrower_throwLocalException cb, Ice.Current current)
    {
        cb.ice_exception(new Ice.TimeoutException());
    }

    public void
    throwNonIceException_async(AMD_Thrower_throwNonIceException cb, Ice.Current current)
    {
        throw new RuntimeException();
    }

    public void
    throwAssertException_async(AMD_Thrower_throwAssertException cb, Ice.Current current)
    {
        throw new java.lang.AssertionError();
    }

    private Ice.ObjectAdapter _adapter;
}

File: js/test/Ice/exceptionsBidir/AMDThrowerI.js
// **********************************************************************
//
// Copyright (c) 2003-2013 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

(function(global){
    var Ice = global.Ice;

    var test = function(b)
    {
        if(!b)
        {
            throw new Error("test failed");
        }
    };

    var AMDThrowerI = function()
    {
    };

    AMDThrowerI.prototype = new TestAMD.Thrower;
    AMDThrowerI.prototype.constructor = AMDThrowerI;

    AMDThrowerI.prototype.shutdown_async = function(cb, current)
    {
        current.adapter.getCommunicator().shutdown();
        cb.ice_response();
    };

    AMDThrowerI.prototype.supportsUndeclaredExceptions_async = function(cb, current)
    {
        cb.ice_response(true);
    };

    AMDThrowerI.prototype.supportsAssertException_async = function(cb, current)
    {
        cb.ice_response(false);
    };

    AMDThrowerI.prototype.throwAasA_async = function(cb, a, current)
    {
        var ex = new TestAMD.A();
        ex.aMem = a;
        cb.ice_exception(ex);
    };

    AMDThrowerI.prototype.throwAorDasAorD_async = function(cb, a, current)
    {
        if(a > 0)
        {
            var ex = new TestAMD.A();
            ex.aMem = a;
            cb.ice_exception(ex);
        }
        else
        {
            var ex = new TestAMD.D();
            ex.dMem = a;
            cb.ice_exception(ex);
        }
    };

    AMDThrowerI.prototype.throwBasA_async = function(cb, a, b, current)
    {
        this.throwBasB(cb, a, b, current);
    };

    AMDThrowerI.prototype.throwBasB_async = function(cb, a, b, current)
    {
        var ex = new TestAMD.B();
        ex.aMem = a;
        ex.bMem = b;
        cb.ice_exception(ex);
    };

    AMDThrowerI.prototype.throwCasA_async = function(cb, a, b, c, current)
    {
        this.throwCasC(cb, a, b, c, current);
    };

    AMDThrowerI.prototype.throwCasB_async = function(cb, a, b, c, current)
    {
        this.throwCasC(cb, a, b, c, current);
    };

    AMDThrowerI.prototype.throwCasC_async = function(cb, a, b, c, current)
    {
        var ex = new TestAMD.C();
        ex.aMem = a;
        ex.bMem = b;
        ex.cMem = c;
        cb.ice_exception(ex);
    };

    AMDThrowerI.prototype.throwUndeclaredA_async = function(cb, a, current)
    {
        var ex = new TestAMD.A();
        ex.aMem = a;
        cb.ice_exception(ex);
    };

    AMDThrowerI.prototype.throwUndeclaredB_async = function(cb, a, b, current)
    {
        var ex = new TestAMD.B();
        ex.aMem = a;
        ex.bMem = b;
        cb.ice_exception(ex);
    };

    AMDThrowerI.prototype.throwUndeclaredC_async = function(cb, a, b, c, current)
    {
        var ex = new TestAMD.C();
        ex.aMem = a;
        ex.bMem = b;
        ex.cMem = c;
        cb.ice_exception(ex);
    };

    AMDThrowerI.prototype.throwLocalException_async = function(cb, current)
    {
        cb.ice_exception(new Ice.TimeoutException());
    };

    AMDThrowerI.prototype.throwLocalExceptionIdempotent_async = function(cb, current)
    {
        cb.ice_exception(new Ice.TimeoutException());
    };

    AMDThrowerI.prototype.throwNonIceException_async = function(cb, current)
    {
        cb.ice_exception(new Error());
    };

    AMDThrowerI.prototype.throwAssertException_async = function(cb, current)
    {
        test(false);
    };

    AMDThrowerI.prototype.throwMemoryLimitException_async = function(cb, seq, current)
    {
        cb.ice_response(Ice.Buffer.createNative(1024 * 20)); // 20KB is over the configured 10KB message size max.
    };

    AMDThrowerI.prototype.throwAfterResponse_async = function(cb, current)
    {
        cb.ice_response();

        throw new Error();
    };

    AMDThrowerI.prototype.throwAfterException_async = function(cb, current)
    {
        cb.ice_exception(new TestAMD.A());

        throw new Error();
    };

    global.AMDThrowerI = AMDThrowerI;
}(typeof (global) === "undefined" ? window : global));
